Creating and Automating DevOps Pipelines for Node-RED in Industrial Environments (#2603)
FlowFuse/website: src/blog/2024/10/how-to-build-automate-devops-pipelines-node-red-deployments.md

---
title: Creating and Automating DevOps Pipelines for Node-RED in Industrial Environments
subtitle: Streamlining Deployments for Efficiency and Safety in Industrial Environments
description: Learn how to build and automate DevOps pipelines for Node-RED deployments in manufacturing and automotive industries. Discover practical tips and tools to streamline your deployment process, ensuring efficiency and safety in critical operations.
date: 2024-10-01
authors: ["sumit-shinde"]
image: 
tags:
   - node-red devops
   - node red devops
   - node red devops pipeline
---

> When deploying a critical application flow update in your manufacturing or automotive process, it is important to perform testing and validation. The smallest errors can lead to a production outage. In such cases, having a quick and reliable deployment process is essential for maximising uptime and minimising downtime.

Reviewer: re suggesting unapplied suggestion from 1st review https://github.com/FlowFuse/website/pull/2603#discussion_r1782892458 

```suggestion
When deploying any update in your manufacturing or automotive process, it is important to perform testing and validation. The smallest errors can lead to a production outage. In such cases, having a quick and reliable deployment process is essential for maximising uptime and minimising downtime.
```

[FlowFuse](/) enhances collaboration, security, and scalability for your Node-RED applications, making the deployment and management of edge devices seamless. With a centralized platform and an intuitive visual interface, FlowFuse allows you to connect, collect, transform, and visualize data effortlessly.

### Steps to Create a DevOps Pipeline:

1. Go to the **FlowFuse platform** and navigate to the application where your Node-RED instances are located. Ensure you have instances set up for all stages, including production devices or instances.
2. Switch to the **DevOps Pipelines** option from the top menu.

![Image showing option to switch to DevOps pipelines tab from top menu](./images/devops-pipeline-option-in-apps.png){data-zoomable}
_Image showing option to switch to DevOps pipelines tab from top menu_

3. Click the **Add Pipeline** button in the top right corner to create the pipeline.

![Screenshot of the FlowFuse dashboard with the 'Add Pipeline' button highlighted at the top-right corner](./images/add-pipeline-button.png){data-zoomable}
_Click the 'Add Pipeline' button to start creating your DevOps pipeline in FlowFuse_

> 4. Once you click **Add Pipeline**, a form opens in the next window. Enter the name for your pipeline and click **Create Pipeline**.

Reviewer: ```suggestion
4.  Enter the name for your pipeline and click **Create Pipeline**.
```

Step 3 already states click the button

_Configure each stage by selecting an instance, device, or device group, and define the deployment actions_

Once you’ve created your initial stage, you can add more stages by following the same process. This flexibility allows you to tailor your DevOps pipeline to meet the specific needs of your Node-RED deployment.

For example, in development, you might have a Node-RED instance in the cloud to build your application. During staging, you could test the setup with a single device. Finally, in production, you can deploy the tested application to thousands of devices in a device group, saving time and ensuring smooth deployment at scale.

![DevOps pipelines animation](./images/devops-pipeline.gif){data-zoomable}
*Image: DevOps animation showing application deployment to production securely.*

## Running a Pipeline Stage

Once your pipeline is set up, you are ready to run it and deploy your changes across each stage. Here's how to do it:

![Screenshot of the 'Run Pipeline' button in FlowFuse, allowing users to trigger the deployment process.](./images/button-to-run-pipeline.png){data-zoomable}

> _Click the 'Run Pipeline' button to initiate the automated deployment process across all configured stages_

Reviewer: ```suggestion
_Click the 'Run Pipeline' button to initiate the a deployment process across for the selected given stage_
```

Original infers all stages will operate

1. Click the run pipeline stage option. For each stage, this button will be available, except for the last one. Simply click it to start the deployment process.

> 2. Once initiated, the deployment will automatically progress from the leftmost stage to the rightmost stage. It will copy all settings, environments, and flows and deploy them to the next stage. However, for environment variables, only the keys will be copied—you will need to manually add the values for each.

Reviewer: This is not accurate. Each 2 adjacent stages operate independently

i.e. stating "the deployment will automatically progress from the leftmost stage to the rightmost stage" infers pressing the leftmost play button will operate the pipeline all the way to the rightmost stage

Additionally, the part about env var keys is misleading.

It goes something like: "The pipeline operation will copy env vars but wont overwrite any that exist in the target."

See this: 

![image](https://github.com/user-attachments/assets/43ff095f-a642-4a4e-8c27-7e7e0e21179a)

description: Learn how to build and automate DevOps pipelines for Node-RED deployments in manufacturing and automotive industries. Discover practical tips and tools to streamline your deployment process, ensuring efficiency and safety in critical operations.
date: 2024-10-01
authors: ["sumit-shinde"]
image: 
tags:
   - node-red devops
   - node red devops
   - node red devops pipeline
---

Imagine deploying a critical software update in your manufacturing or automotive facility. Without a proper pipeline that includes various stages of testing and validation, a small error could lead to a floor crash or a production halt. In these industries, having a quick and reliable deployment process is essential for maintaining efficiency and safety.

For developers using Node-RED, setting up a comprehensive DevOps pipeline can make all the difference. In this blog, we’ll explore how to build and automate DevOps pipelines specifically for Node-RED deployments. You’ll discover practical tips and tools to streamline your process, ensuring your applications are always ready to support your operations.

## What Exactly is a DevOps Pipeline?

> 

Reviewer 2: An image would be great here @sumitshinde-84 (cc @Yndira-E)

A DevOps pipeline is an automated process that helps developers move their code from development to production smoothly. In a pipeline, each step depends on the one before it, ensuring every update is properly tested and ready for use. The main stages are:

1. **Development:** This is where developers build and test their application. Testing ensures that basic issues are caught early.
2. **Staging/Testing:** In this stage, the code is deployed to a staging environment that closely mimics the live system. Here, the application is tested in real-world conditions to catch any remaining bugs or issues.
3. **Production:** If everything works fine in staging, the code is deployed to production, making it live and accessible to users.

## How to Create DevOps Pipelines for Node-RED Deployments?

Creating DevOps pipelines manually for your Node-RED deployments can be time-consuming, expensive, and require considerable technical expertise. **FlowFuse** simplifies the creation of DevOps pipelines for Node-RED deployments.

<div class="blog-update-notes">

>     <p>[FlowFuse](/) enhances collaboration, security, and scalability for your Node-RED applications, making the deployment and management of edge devices seamless. With a centralized platform and an intuitive visual interface, FlowFuse allows you to connect, collect, transform, and visualize data effortlessly.</p>

Reviewer 2: > [FlowFuse](/)

This renders raw in the blog article. Assume you're meaning it to be a link?

Odd choice to have it highlighted this way too, it reads as part of the article, but the styling makes it feel like an update notice or something like that

Author: Removing that highlight.

</div>

### Steps to Create a DevOps Pipeline:

1. Go to the **FlowFuse platform** and navigate to the application where your Node-RED instances are located. Ensure you have instances set up for all stages, including production devices or instances.

> ![Image showing option to switch to DevOps pipelines tab from top menu](./images/devops-pipeline-option-in-apps.png){data-zoomable}

Reviewer 2: This image belongs to Step 2, but appears above it, generally it should be written step _first_, then the image/screenshot.

This is also the case for every screenshot used here in these steps.

6. In the window that opens, select the **stage type** based on whether it's an instance, device, or device group.
7. Enter the name for the stage in the **Stage Name** field.
8. Choose an instance, device, or device group for the stage.
9. Next, configure which action should be performed when this stage is pushed to the next:
   - **Create New Snapshot:** Generates a new [Snapshot](/docs/user/high-availability/) using the current flows and settings.
   - **Use Latest Instance Snapshot:** Uses the most recent existing snapshot of the instance. The deployment will fail if no snapshot exists.
   - **Prompt to Select Snapshot:** Prompts at deploy time to select which snapshot from the source stage should be copied to the next stage.
10. Check the option **Deploy to Devices** if you want changes to be deployed to all devices connected to this stage’s instance when the stage is deployed.

Once you’ve created your initial stage, you can add more stages by following the same process. This flexibility allows you to tailor your DevOps pipeline to meet the specific needs of your Node-RED deployment.

> ![DevOps pipelines animation](./images/devops-pipeline.gif){data-zoomable}

Reviewer 2: Broken image that doesn't render

description: Learn how to build and automate DevOps pipelines for Node-RED deployments in manufacturing and automotive industries. Discover practical tips and tools to streamline your deployment process, ensuring efficiency and safety in critical operations.
date: 2024-10-03
authors: ["sumit-shinde"]
image: /blog/2024/10/images/creating-and-automating-devops-pipeline-nr.png
tags:
   - node-red devops
   - node red devops
   - node red devops pipeline
---

When deploying any update in your manufacturing or automotive process, it is important to perform testing and validation. The smallest errors can lead to a production outage. In such cases, having a quick and reliable deployment process is essential for maximising uptime and minimising downtime.
<!--more-->
For developers using Node-RED, setting up a comprehensive DevOps pipeline can make all the difference. In this blog, we’ll explore how to build and automate DevOps pipelines specifically for Node-RED deployments. You’ll discover practical tips and tools to streamline your process, ensuring your applications are always ready to support your operations.

## What Exactly is a DevOps Pipeline?

> ![Image showing the DevOps Lifecycle](./images/What-is-DevOps.png){data-zoomable}

Reviewer 3: @sumitshinde-84 I’ve replaced the one you added because it was pixelated. I found another one with higher resolution.

> _The DevOps Lifecycle illustrates the stages of development, testing, and deployment, ensuring a smooth transition from development to production for efficient and reliable software delivery_

Reviewer: ```suggestion
_The DevOps Lifecycle illustrating the stages of development through to production_
```

Is there really any need for such lengthy titles when the whole article is telling the reader what devops is?

   - **Prompt to Select Snapshot:** Prompts at deploy time to select which snapshot from the source stage should be copied to the next stage.
10. Check the option **Deploy to Devices** if you want changes to be deployed to all devices connected to this stage’s instance when the stage is deployed.

![Screenshot showing the stage configuration form with options to select instance types and configure actions.](./images/form-to-create-configure-stages.png){data-zoomable}
_Configure each stage by selecting an instance, device, or device group, and define the deployment actions_

Once you’ve created your initial stage, you can add more stages by following the same process. This flexibility allows you to tailor your DevOps pipeline to meet the specific needs of your Node-RED deployment.

For example, in development, you might have a Node-RED instance in the cloud to build your application. During staging, you could test the setup with a single device. Finally, in production, you can deploy the tested application to thousands of devices in a device group, saving time and ensuring smooth deployment at scale.

![DevOps pipelines animation](./images/devops-pipeline.gif){data-zoomable}

> *Image: DevOps animation showing application deployment to production securely.*

Reviewer: ```suggestion
*Image: DevOps animation demonstrating pipeline deployments.*
```

SUGGESTION - Not sure why you felt the need to state "securely?"

## Running a Pipeline Stage

Once your pipeline is set up, you are ready to run it and deploy your changes across each stage. Here's how to do it:

![Screenshot of the 'Run Pipeline' button in FlowFuse, allowing users to trigger the deployment process.](./images/button-to-run-pipeline.png){data-zoomable}

> _Click the 'Run Pipeline' button to initiate the a deployment process across for the selected given stage_

Reviewer: ```suggestion
_Click the 'Run Pipeline' button to initiate a pipeline deployment_
```

Sentence still not really making sense.  KISS ;)

1. Click the run pipeline stage option of first stage. For each stage, this button will be available, except for the last one. Simply click it to start the deployment process.
2. Once initiated, the deployment will automatically progress from the current stage to the next stage on the right. For example, if you have three stages—Development, Staging, and Production—clicking the "Run Pipeline" button for the Development stage will advance the deployment to the Staging stage. To proceed from Staging to Production, you will need to click the "Run" button for the Staging stage, as each pair of adjacent stages operates independently.

> This action will create a new snapshot in the current stage and copy all settings, environment variables, and flows to the next stage, but it will not overwrite any existing environment variable keys in the target stage.

Reviewer: This section is now far too wordy and confusing.

`1` starts by talking about the "first stage" then goes on to mention "each stage"
`2` continues on as if you never mentioned "each stage" in `1`)
the example added to `2` is very long and detracts from the simple steps

the following sentence state "This action will create a new snapshot in the current stage" but i am lost as to which stage you are referring to.

> It’s essential to remember that each stage acts as either a source or a target: stages (except the last) push forward, while stages (except the first) receive results from the previous one. This ensures an organized, left-to-right flow of deployment without overlooked steps.

Reviewer: > It’s essential to remember that each stage acts as either a source or a target: stages (except the last) push forward, while stages (except the first) receive results from the previous one.

I think this can (should) be simplified.  
I am unsure while it is necessary to say "It’s essential to remember". 
Then is says "acts as either a source or a target:" follow by 2 exceptions "(except the last)" and "(except the first)"
Can we think of a cleaner way to state this?
Do we need this sentence at all? I think we have already demonstrated and explained the pipeline operates from left to right!

NOTE: if you do remove this sentence, the next sentence may need modifying as it starts with "Also"

Once your pipeline is set up, you are ready to run it and deploy your changes across each stage. Here's how to do it:

![Screenshot of the 'Run Pipeline' button in FlowFuse, allowing users to trigger the deployment process.](./images/button-to-run-pipeline.png){data-zoomable}

> _Click the 'Run Pipeline' button to initiate the a deployment process across for the selected given stage_

Reviewer: This sentence is not well formed. Please double check it.
